### PR TITLE
Skip compiling fusion segments when using expression evaluator

### PR DESCRIPTION
**Pull request by Priya2698**
This PR resolves Issue #1812.
1. Removes the workarounds added to support ATen evaluation for matmuls in PR #1775 (in `predicate_elimination.cpp` and `validation.cpp` (see comments in the PR)
2. `FusionExecutor::fusion_` is initialized only when compilation is skipped. So only one of `lowered_` or `fusion_` is non-null.
3. `FusionKernelRuntime/FusionExecutorCache/FusionExecutor::isCompiled()` indicates whether compilation was attempted (either marked for EE or via nvFuser).

### CODE REVIEW

**Comment by Priya2698** [comment]
!build

**Comment by Priya2698** [comment]
!build

**Comment by Priya2698** [comment]
!build

**Comment by Priya2698** [comment]
!build

**Comment by Priya2698** [comment]
!build

**Comment by Priya2698** [comment]
!build

**Comment by Priya2698** [comment]
!build

File: csrc/device_lower/analysis/predicate_elimination.cpp [hunk 1]
```diff
@@ -890,9 +890,7 @@ class PredicateChcker : public IterVisitor {
 } // namespace
 
 PredicateElimination::PredicateElimination(Fusion* fusion) {
```
Comment (Priya2698) [inline csrc/device_lower/analysis/predicate_elimination.cpp:892]: This was a WAR. Removing now, since we are skipping compilation.

File: csrc/device_lower/validation.cpp [hunk 2]
```diff
@@ -1030,12 +1030,7 @@ void validateSizeMemoryOp(LoadStoreOp* ldst) {
 //! Validate data format and GPU arch compatibility of scheduled
 //!  mma operators on the fusion.
 void validateMma(Fusion* fusion) {
```
Comment (Priya2698) [inline csrc/device_lower/validation.cpp:1032]: This was a WAR. Removing now, since we are skipping compilation.

File: csrc/executor.h [hunk 3]
```diff
@@ -199,6 +205,11 @@ class FusionExecutor : public NonCopyable {
     return lowered_->kernel();
   }
 
+  Fusion* fusion() const {
+    NVF_ERROR(lowered_ || fusion_);
```
Comment (wujingyue) [inline csrc/executor.h:209]:
```suggestion
    NVF_ERROR(lowered_ ^ fusion_);
```

Xor or Or?

File: csrc/serde/fusion_cache.fbs [hunk 4]
```diff
@@ -345,6 +345,7 @@ table CudaKernel {
 
 // Each Fusion Executor maps to a lowered and compiled kernel.
 table FusionExecutor {
+  is_compilation_skipped : bool;
```
Comment (wujingyue) [inline csrc/serde/fusion_cache.fbs:348]: Why is this needed? isCompilationSkipped is a method that describes the state of the FusionExecutor. Making it a field in `table FusionExecutor` seems to suggest it's part of the state itself. cc @rdspring1

File: csrc/kernel_cache.cpp [hunk 5]
```diff
@@ -1379,7 +1380,9 @@ std::pair<LaunchParams, CompileParams> FusionKernelRuntime::getKernelConfig(
 
   // Check that the heuristics are matched, in the case of segmented fusion
   NVF_ERROR(!sg || scheduler_entry->heuristic() == sg->heuristic());
-  NVF_ERROR(executors_.at(group_id).isCompiled());
+  NVF_ERROR(
+      executors_.at(group_id).isCompiled() ||
+      executors_.at(group_id).isCompilationSkipped());
```
Comment (wujingyue) [inline csrc/kernel_cache.cpp:1385]: Why does this check exist in the first place? The function itself doesn't seems to use `fusion_` or `kernel_`. It seems to belong to something like FusionExecutor::runFusion.

File: tests/cpp/test_resize.cpp [hunk 6]
```diff
@@ -1900,6 +1902,9 @@ TEST_F(ResizeTest, FusionSliceForNanoGPT2) {
 
 // C++ version of TestNvFuserFrontend.test_nanogpt_split_mha_linears
 TEST_F(ResizeTest, FusionSliceForNanoGPT3) {
+  preseg_passes::OptimizationPassGuard<preseg_passes::MarkAliasesPreparePass>
+      optimization_guard(false);
```
Comment (jjsjann123) [inline tests/cpp/test_resize.cpp:1906]: naive question, why are we skipping alias pass on this?

File: tests/cpp/test_gpu3.cpp [hunk 7]
```diff
@@ -31,6 +31,8 @@
 #include <kernel_ir.h>
 #include <kernel_ir_dispatch.h>
 #include <ops/all_ops.h>
+#include <preseg_passes/mark_aliases_prepare.h>
+#include <preseg_passes/optimization_pass.h>
```
Comment (jjsjann123) [inline tests/cpp/test_gpu3.cpp:35]: nitpick: Don't see any reason to add these headers.

File: csrc/kernel_cache.h [hunk 8]
```diff
@@ -124,7 +124,7 @@ class FusionKernelRuntime {
     std::lock_guard<std::mutex> guard(mutex_);
     return std::all_of(
         executors_.begin(), executors_.end(), [](const auto& executor) {
-          return executor.isCompiled();
+          return executor.isCompiled() || executor.isCompilationSkipped();
```
Comment (jjsjann123) [inline csrc/kernel_cache.h:127]: nitpick: This feels a little bit confusing to me.

i.e. `FusionKernelRuntime::isCompiled` means something different from `FusionExecutor::isCompiled`. (similarly we have `FusionExecutorCache::isCompiled`.

We don't have to do it in this PR, but maybe another renaming PR to clean up the `isCompiled` vs `isCompilationSkipped` in `FusionExecutor`.

File: csrc/executor.h [hunk 9]
```diff
@@ -398,6 +409,14 @@ class FusionExecutor : public NonCopyable {
       int64_t runtime_id,
       int64_t group_id);
 
+  //! Check if compilation was skipped (fusion segment marked for EE).
+  bool isCompilationSkipped() const {
+    if (fusion_ != nullptr) {
+      NVF_ERROR(!lowered_, "Expected GPU lowering to be skipped.");
+    }
```
Comment (jjsjann123) [inline csrc/executor.h:416]:
```suggestion
    NVF_ERROR(fusion_ == nullptr || !lowered_, "Expected GPU lowering to be skipped.");
```

nitpicking is all I do.

File: csrc/kernel_cache.cpp (hunk 5 above)
Comment (Priya2698) [reply csrc/kernel_cache.cpp:1385]: I only see `scheduler_entry->params()->cparams.index_type` being updated in `compileFusion`.
Since this function is only called here before `runFusion`, we can move the checks inside `runFusion`.

Only if this function is used in isolation (which should probably not happen) it can have `index_type` vary based on compilation status

File: tests/cpp/test_resize.cpp (hunk 6 above)
Comment (Priya2698) [reply tests/cpp/test_resize.cpp:1906]: I'll add a comment to explain.

This test is checking 
```
NVF_CHECK(
      !kernel->summary().has_cooperative_grid_reduction,
      "Grid sync should not be used as slicing input should avoid input caching");
```
If we don't skip aliasing, the compilation will be skipped, and this property cannot be verified.
So I am skipping this pass, so the test can actually verify this behavior.

File: tests/cpp/test_gpu3.cpp (hunk 7 above)
Comment (Priya2698) [reply tests/cpp/test_gpu3.cpp:35]: Right, this is a remnant from previous changes. Thanks for catching!

File: csrc/kernel_cache.h (hunk 8 above)
Comment (Priya2698) [reply csrc/kernel_cache.h:127]: I should rename it to `FusionKernelRuntime::isCompiledOrCompilationSkipped`.
This function is used to check if compilation was already attempted which is true if either the executors are all compiled or compilation was skipped for them.

Comment (Priya2698) [reply csrc/kernel_cache.h:127]: Any other naming suggestions for this function?

File: tests/cpp/test_resize.cpp (hunk 6 above)
Comment (Priya2698) [reply tests/cpp/test_resize.cpp:1906]: We could also modify the test itself, to have other operators to avoid the outputs being aliased.

File: csrc/kernel_cache.h (hunk 8 above)
Comment (rdspring1) [reply csrc/kernel_cache.h:127]: 1. isCompilationSkipped() --- Returns True when we use the Aten function and `FusionExecutor` is never compiled.
2. isCompiled() --- Returns False when `FusionExecutor` has yet to be compiled.

Maybe we should rename `isCompilationSkipped()` to `isAtenExecuted()` or `isEagerModeExecutor()`

Comment (Priya2698) [reply csrc/kernel_cache.h:127]: I agree with Ryan's suggestion. We can keep `isCompiled` to denote if compilation was attempted -- either a kernel was compiled or we are using Aten evaluation. Additionally, `isCompilationSkipped` is renamed to `isEagerModeExecutor`.

In this naming scheme, `isCompiled` always means whether or not we attempted to compile a fusion.

Wdyt @jjsjann123?

Comment (rdspring1) [reply csrc/kernel_cache.h:127]: When `isEagerModeExecutor` is `false`, we're using the nvfuser executor.

`isCompiled` implies that we have a cubin for the given fusion. There is an implicit assumption that the nvfuser executor is being used. 

I'd consider renaming `isCompiled` to `isCubinCompiled`, but leaving it alone is also fine.

Comment (Priya2698) [reply csrc/kernel_cache.h:127]: Okay, so the renaming is 
1. `FusionExecutor::isCompiled` -> `FusionExecutor::isCubinCompiled`
2. `FusionExecutor::isCompilationSkipped` -> `FusionExecutor::isEagerModeExecutor`
3. `FusionKernelRuntime/FusionExecutorCache::isCompiled` -> no change. Returns true if either of the above are true indicating that compilation was attempted.

I'll go with this naming if it clears the confusion. @rdspring1 @jjsjann123

File: csrc/serde/fusion_cache.fbs (hunk 4 above)
Comment (wujingyue) [reply csrc/serde/fusion_cache.fbs:348]: Sorry for the delay -- I missed the notification. 

I understood how it's used [here](https://github.com/NVIDIA/Fuser/pull/1930/files#diff-e72d395ea8a3ab3d2a909dfb3677beac1bbd89336b54ee858867f652b48be4f0R2339) to avoid unnecessary compilation. However, can we replace that check with a check on the fusion itself instead of introducing yet another flag? If works, this can avoid an additional state that has to be consistent with some existing state of the fusion.

File: csrc/kernel_cache.h (hunk 8 above)
Comment (jjsjann123) [reply csrc/kernel_cache.h:127]: sounds very reasonable to me.

**Comment by Priya2698** [comment]
!build

**Comment by Priya2698** [comment]
Tests pass here: https://gitlab-master.nvidia.com/dl/pytorch/fuser-gh-mirror/-/pipelines/14499669

File: csrc/executor.h [hunk 10]
```diff
@@ -164,11 +170,26 @@ class FusionExecutor : public NonCopyable {
...
+    NVF_ERROR(
+        lowered_ == nullptr || !fusion_,
+        "fusion_ should only be initializaed when using expression evaluator.");
+    return fusion_ || lowered_;
+  };
+
   // function to query whether a `FusionExecutor` has a compiled kernel to
   // execute
-  bool isCompiled() const {
+  bool hasCompiledKernel() const {
```
Comment (jjsjann123) [inline csrc/executor.h:187]: :+1:

File: csrc/executor.h [hunk 11]
```diff
@@ -164,11 +170,26 @@ class FusionExecutor : public NonCopyable {
     post_lowering_hooks_.push_back(std::move(hook));
   }
 
+  // Function to query whether compilation was attempted for a `FusionExecutor`
+  bool isCompiled() const {
```
Comment (jjsjann123) [inline csrc/executor.h:174]: looks like we changed the plan of renaming here.

File: csrc/executor.h [hunk 12]
```diff
@@ -164,11 +170,26 @@ class FusionExecutor : public NonCopyable {
...
+    // Check atmost one of fusion_ and lowered_ is null.
```
Comment (jjsjann123) [inline csrc/executor.h:175]:
```suggestion
    // Check at most one of fusion_ and lowered_ is null.
```

File: csrc/executor.h [hunk 13]
```diff
@@ -164,11 +170,26 @@ class FusionExecutor : public NonCopyable {
...
+    NVF_ERROR(
+        fusion_ == nullptr || !lowered_,
+        "Expected GPU lowering to be skipped.");
+    NVF_ERROR(
+        lowered_ == nullptr || !fusion_,
+        "fusion_ should only be initializaed when using expression evaluator.");
```
Comment (jjsjann123) [inline csrc/executor.h:181]: out of curiosity, `!x` is just to check `x != nullptr`, right?

I don't see how the above check is associated with the error message. We are only checking that there's one initialized at most, so that's a `!(fusion_ && lowered_)`

File: csrc/executor.h (hunk 11 above)
Comment (jjsjann123) [reply csrc/executor.h:174]: I'm not nitpicking on this BTW.

Comment (Priya2698) [reply csrc/executor.h:174]: Yes, I decided to keep `isCompiled` the same at every place. At `FusionExecutorCache / FusionKernelRuntime` level, we should not need to distinguish between nvFuser vs ATen evaluation generally, so trying to limit that detail within `FusionExecutor`

File: csrc/executor.h [hunk 14]
```diff
@@ -199,6 +220,11 @@ class FusionExecutor : public NonCopyable {
     return lowered_->kernel();
   }
 
+  Fusion* fusion() const {
+    NVF_ERROR((lowered_ && !fusion_) || (!lowered_ && fusion_));
```
[the remark this answers is not in the document]
Comment (Priya2698) [reply csrc/executor.h:224]: Yes, but XOR does not work on these objects.

[the remark this answers is not in the document]
Comment (jjsjann123) [reply csrc/executor.h]: TIL. `explicit operator bool()`

**Review by jjsjann123 — APPROVED** [review]
Thanks for addressing my concerns.

**Comment by Priya2698** [comment]
!build

File: csrc/executor.h [hunk 15]
```diff
@@ -199,6 +215,13 @@ class FusionExecutor : public NonCopyable {
     return lowered_->kernel();
   }
 
+  Fusion* fusion() const {
+    NVF_ERROR(
+        (lowered_ && !fusion_) || (!lowered_ && fusion_),
```
Comment (wujingyue) [inline csrc/executor.h:220]: I think it's clearer to say 
```
(lowered_ == nullptr) != (fusion_ == nullptr)
```
which essentially use `!=` to do a logical xor.

File: csrc/executor.h [hunk 16]
```diff
@@ -398,6 +421,14 @@ class FusionExecutor : public NonCopyable {
       int64_t runtime_id,
       int64_t group_id);
 
+  //! Check if compilation was skipped (fusion segment marked for EE).
+  bool isExprEval() const {
```
Comment (wujingyue) [inline csrc/executor.h:425]: How about isExpressionEvaluated for less ambiguity?

File: tests/cpp/test_resize.cpp [hunk 17]
```diff
@@ -1900,6 +1902,11 @@ TEST_F(ResizeTest, FusionSliceForNanoGPT2) {
 
 // C++ version of TestNvFuserFrontend.test_nanogpt_split_mha_linears
 TEST_F(ResizeTest, FusionSliceForNanoGPT3) {
+  // To verify input caching condition in this test, disable aliasing as that
+  // will skip compilation and no kernel will exist.
+  preseg_passes::OptimizationPassGuard<preseg_passes::MarkAliasesPreparePass>
+      optimization_guard(false);
+
   auto fusion_ptr = std::make_unique<Fusion>();
```
Comment (wujingyue) [inline tests/cpp/test_resize.cpp:1910]: I know you didn't add this, but since you are here, do you mind doing a minor, side refactor? 

1. Change this line to
   ```
   auto fusion = std::make_unique<Fusion>();
   ```
2. Remove the next line says `auto& fusion=*fusion_ptr`. 
3. Change existing uses of `fusion.` to `fusion->`. 

The gist is that `unique_ptr` comes with syntax sugar to make it as if it's a raw pointer. It's not necessary to create another variable just to hold its reference.
